src/analyses/call_graph.cpp

```diff
@@ -443,3 +491,91 @@ void find_leaves_below_function(
   std::unordered_set<irep_idt, dstring_hash> to_avoid;
   find_leaves_below_function(call_graph, function, to_avoid, output);
 }
+
+/*******************************************************************\
+
+Function: find_direct_or_indirect_callees_of_function
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void find_direct_or_indirect_callees_of_function(
+  const call_grapht &call_graph,
+  const irep_idt &function,
+  std::unordered_set<irep_idt, dstring_hash> &output)
+{
+  std::unordered_set<irep_idt, dstring_hash> leaves;
+  find_leaves_below_function(call_graph, function, output, leaves);
+  output.insert(leaves.cbegin(), leaves.cend());
+}
+
+/*******************************************************************\
+
+Function: find_nearest_common_callees
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void find_nearest_common_callees(
```

Comment:
Couple of years ago I read in one of Scott Mayer's books the following recommendation for clean OO design in C++: "When an algorithm can be effectively implemented using only the public interface on a class, then the algorithm should be implemented as a free function."

It makes sense to me (i.e. the intention to minimise a set of algorithms with the direct access to private data of the class maximises the encapsulation of that class, and so it simplifies definition and maintenance of class invariants), so I follow it. However, one can easily disagree. For example, it would be quite difficult to follow this recommendation in Java.

I am not sure whether you only asked the question or whether you actually want me to make it a member function. So, please, let me know...